File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    bool is_valid_helper(TreeNode* root, int min, int max) {
        if (root == nullptr) return true;
        if (root->val >= max || root->val <= min) {
            return false;
        }
        return is_valid_helper(root->left, min, root->val) &&
               is_valid_helper(root->right, root->val, max);
    }
    bool isValidBST(TreeNode* root) { return is_valid_helper(root, INT_MIN, INT_MAX); }
};

File: Data Structures & Algorithms/populating-next-right-pointers-in-each-node/submission-5.js
/**
 * Definition for a binary tree node.
 * class Node {
 *     constructor(val = 0, left = null, right = null, next = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {Node} root
     * @return {Node}
     */
    connect(root) {
        function dfs(root) {
            if (!root) return;
            if (root.left && root.right) {
                root.left.next = root.right;
                if (root.next) {
                    root.right.next = root.next.left;
                }
            } 
            dfs(root.left);
            dfs(root.right);
        }
        dfs(root);
        return root;
    }
}
